nova/virt/vmwareapi/vmops.py

```diff
@@ -3113,15 +3111,32 @@ def _attach_volumes(self, instance, block_device_info, adapter_type,
                         existing_disks=None):
         disks = driver.block_device_info_get_mapping(block_device_info)
         # make sure the disks are attached by the device_name order
-        for disk in sorted(disks,
-                           key=itemgetter('mount_device')):
+        for disk in self._sort_disks(disks):
             if existing_disks and disk['volume_id'] in existing_disks:
                 continue
 
             adapter_type = disk.get('disk_bus') or adapter_type
             self._volumeops.attach_volume(disk['connection_info'], instance,
                                           adapter_type)
 
+    def _sort_disks(self, disks, reverse=False):
+        """Since mount_device can be controlled from outside, we want to
+        make sure the boot_index >= 0 devices are always at the top.
+        """
+        boot_disks = sorted(
+            (d for d in disks if d.get('boot_index') is not None),
+            key=itemgetter('boot_index'),
+            reverse=reverse)
```

Comment:
This would put `boot_index` set to `-1` before `0`, but `0` is the root disk afaik. [This comment](https://github.com/sapcc/nova/blob/b13c04d577cc4d82487c6f95cc1efd690d5458ca/nova/block_device.py#L200-L201) states that `-1` is the same as `None`.

During spawn, we already sort by `boot_index != 0` [code](https://github.com/sapcc/nova/blob/b13c04d577cc4d82487c6f95cc1efd690d5458ca/nova/virt/vmwareapi/vmops.py#L1600-L1601).

I think we could sort by `(d.get('boot_index') != 0, d.get('mount_device')`